docs/function.md

```diff
@@ -110,231 +108,144 @@ let eleven = addFive(6);
+makeCircle(~radius=1);
 ```
 
-Is `whatIsThis` a curried `drawCircle` function, waiting for the optional `radius` to be applied? Or did it finish applying because the `radius` is optional? To address this confusion, append a positional (aka non-labeled) argument to `drawCircle` (conventionally `()`), and OCaml will, as a rule of thumb, presume the optional labeled argument is omitted when the positional argument is provided.
+It is not clear if that should return a function waiting for `x` and `y`
+coordinates, or if should default the `x` and `y` coordinates to zeroes. With
+the final unit argument these cases are disambiguated:
 
-If we don't supply the unit, OCaml knows we want to curry the function.
 ```reason
-let curriedFunction = drawCircle(~color);
-```
+/* Function waiting for x and y coordinates. */
+let makeUnitCircle = makeCircle(~radius=1);
+let c = makeUnitCircle(~x=5, ~y=5, ());
 
-If we _do_ supply the unit, OCaml knows we deliberately omit the `radius` parameter, and the function is executed.
-```reason
-let circle = drawCircle(~color, ());
+/* Creates a circle at (0, 0) with radius 1 */
+let c = makeCircle(~radius=1, ());
 ```
 
-### Explicitly Passed Optional
+### Optional Arguments
 
-Sometimes, you might want to forward a value to a function without knowing whether the value is `None` or `Some(a)`. Naively, you'd do:
+Similar to default argument values, named arguments can be marked as optional.
+This is just syntax around making the default value `None`.
 
 ```reason
-let result =
-  switch (payloadRadius) {
-  | None => drawCircle(~color, ())
-  | Some(r) => drawCircle(~color, ~radius=r, ())
+let next = (~value=?, ()) => {
+  switch (value) {
+  | Some(value) => value + 1
+  | None => 1
   };
-```
-
-This quickly gets tedious. We provide a shortcut:
-
-```reason
-let result = drawCircle(~color, ~radius=?payloadRadius, ());
-```
-
-This means "I understand `radius` is optional, and that when I pass it a value it needs to be an `int`, but I don't know whether the value I'm passing is `None` or `Some(val)`, so I'll pass you the whole `option` wrapper".
-
-### Optional with Default Value
-
-Optional labeled arguments can also be provided a default value. In this case, they aren't wrapped in an `option` type.
-
-```reason
-let drawCircle = (~radius=1, ~color, ()) => {
-  setColor(color);
-  startAt(radius, radius)
 };
-```
-
-### Recursive Functions
-
-By default, a value can't see a binding that points to it, but including the `rec` keyword in a `let` binding makes this possible. This allows functions to see and call themselves, giving us the power of recursion.
 
-```reason
-let rec neverTerminate = () => neverTerminate();
+next(); /* 1 */
+next(~value=11, ()); /* 12 */
 ```
 
-### Mutually Recursive Functions
-
-Mutually recursive functions start like a single recursive function using the
-`rec` keyword, and then are chained together with `and`:
+Sometimes you will have an option that needs to be passed to a function with an
```

Comment:
I realized that the terminology of "optional argument" is confusing because it has the word "option" in it. That might be the main source of confusion I see from people with respect to named arguments with defaults. That immediately makes people think that the type they pass to the function is of type `option<t>`, when it's not the case.

What could fix that confusion by using the terminology "defaulted argument". A function that accepts a `~name` argument that is defaulted, _still_ accepts a `string`. It just happens to be defaulted to something if it wasn't supplied by the caller.

Then this section could be much less confusing if described like:
> Sometimes you will have a function that specifies a named argument with a defaulted value. The caller can supply that argument, or omit it. It is common to decide whether or not to supply that argument based on whether or not some option variable is `None` vs. `Some(data)`.
> ```reason
> switch(myVariable) {
> | Some(data) => callFunction(~data=data, 100)
> | None => callFunction(100)
> }
> ```
> Since this is a common pattern, Reason provides a shorthand syntax for the pattern above:
> Prefixing the named argument value with a question mark, will cause that named argument to
> be supplied with the contents of `Some(contents)` if the value is indeed `Some(..)`, otherwise
> the named argument will not be supplied at all.
> ```reason
> callFunction(~data=?data, 100)
> }
> ```